### ISSUE TITLE
[Feature] Is it possible to replace linear_sum_assignment of scipy with the solution from this repo

### ISSUE THREAD

**Issue by henryzhengr** [ivan-chai/torch-linear-assignment#3]
Hi author,

Would like to ask if it is possible to replace linear_sum_assignment with code from this repo?

For example, I tried:
`cost = torch.tensor([
    8, 4, 7,
    5, 2, 3,
    9, 6, 7,
    9, 4, 8
]).reshape(1, 4, 3)`

Function 1: `workers, tasks = linear_sum_assignment(cost[i].numpy())  # (N, 2).`
Output: `print(workers, tasks) #(tensor([0, 1, 3]), tensor([0, 2, 1]))`

Function 2:     
```
if t < w:
        cost = cost.transpose(1, 2)  # (B, T, W).
        col4row, row4col = backend.batch_linear_assignment(cost.contiguous())  # (B, T), (B, W).
else:
        col4row, row4col = backend.batch_linear_assignment(cost.contiguous())  # (B, W), (B, T).
```
Output: `print(col4row.cpu(), row4col.cpu()) #(tensor([[0, 3, 1]], dtype=torch.int32), tensor([[ 0,  2, -1,  1]], dtype=torch.int32))`

The output seems to be different, can we replace the linear_sum_assignment from scipy with your solution? Looking forward to your reply

**Comment by ivan-chai** [comment]
Hi,

It's hard to make interface compatible with Scipy. The Scipy's result depends on the number of assignments. In your example it is equal to 3. But for other tasks with the same shape of the cost matrix, the output can be 0, 1, 2, 3 or 4 pairs, and we can't gather results in one tensor of shape (B, W, 2) or (B, T, 2).

However, we can compute the desired output like that:
```
cost = torch.tensor([ 8, 4, 7, 5, 2, 3, 9, 6, 7, 9, 4, 8 ]).reshape(1, 4, 3)
assignment = batch_linear_assignment(cost.cuda())  # (B, W).
assignment = assignment[0]  # (W).
print(assignment)  # tensor([ 0,  2, -1,  1], device='cuda:0')

mask = assignment >= 0
workers = mask.nonzero().flatten()
print(workers)  # tensor([0, 1, 3], device='cuda:0')

tasks = assignment[mask].argsort()
print(tasks)  # tensor([0, 2, 1], device='cuda:0')
```

**Comment by ivan-chai** [comment]
Think, we can add interface similar to Scipy in the next release.

**Comment by henryzhengr** [comment]
Thank you! This solution works!

**Comment by PYNing** [comment]
> Hi,
> 
> Thank you for your suggestion!
> 
> We can compute the desired output for each element of the batch like that:
> 
> ```
> cost = torch.tensor([ 8, 4, 7, 5, 2, 3, 9, 6, 7, 9, 4, 8 ]).reshape(1, 4, 3)
> assignment = batch_linear_assignment(cost.cuda())  # (B, W).
> assignment = assignment[0]  # (W).
> print(assignment)  # tensor([ 0,  2, -1,  1], device='cuda:0')
> 
> mask = assignment >= 0
> workers = mask.nonzero().flatten()
> print(workers)  # tensor([0, 1, 3], device='cuda:0')
> 
> tasks = assignment[mask].argsort()
> print(tasks)  # tensor([0, 2, 1], device='cuda:0')
> ```

Thanks for your work, but I have to point out that the code above has a bug with the test case：
cost = torch.randn((bs, rows, cols))，which bs=1, rows=2900, cols=200

`tasks` in `linear_sum_assignment ` is **NOT** equal to `assignment[mask].argsort()` ，while the right code should be `row4col.argsort()`('row4col' means the second output of `backend.batch_linear_assignment()`).

Therefore the interface still needs to be refactored for this feature.

**Comment by ivan-chai** [comment]
Hi! Thank you for pointing out this issue. Indeed, there is a problem with this code: argsort must be removed.

Try this code:

```
import torch
import numpy as np
from torch_linear_assignment import batch_linear_assignment
from scipy.optimize import linear_sum_assignment

cost = torch.randn(1, 2900, 200).cuda()
assignment = batch_linear_assignment(cost)  # (B, W).
assignment = assignment[0]  # (W).

mask = assignment >= 0
workers = mask.nonzero().flatten()
tasks = assignment[mask]   # Without argsort!

sc_workers, sc_tasks = linear_sum_assignment(cost[0].cpu().double())
sc_workers = torch.tensor(sc_workers).cuda()
sc_tasks = torch.tensor(sc_tasks).cuda()
print((workers ==  sc_workers).all(), (tasks == sc_tasks).all())
```

**Comment by PYNing** [comment]
> Hi! Thank you for pointing out this issue. Indeed, there is a problem with this code: argsort must be removed.
> 
> Try this code:
> 
> ```
> import torch
> import numpy as np
> from torch_linear_assignment import batch_linear_assignment
> from scipy.optimize import linear_sum_assignment
> 
> cost = torch.randn(1, 2900, 200).cuda()
> assignment = batch_linear_assignment(cost)  # (B, W).
> assignment = assignment[0]  # (W).
> 
> mask = assignment >= 0
> workers = mask.nonzero().flatten()
> tasks = assignment[mask]   # Without argsort!
> 
> sc_workers, sc_tasks = linear_sum_assignment(cost[0].cpu().double())
> sc_workers = torch.tensor(sc_workers).cuda()
> sc_tasks = torch.tensor(sc_tasks).cuda()
> print((workers ==  sc_workers).all(), (tasks == sc_tasks).all())
> ```

It works, THX.

**Comment by ivan-chai** [comment]
We have added compatibility with SciPy in the batch mode:

https://github.com/ivan-chai/torch-linear-assignment/blob/main/torch_linear_assignment/assignment.py#L54

The example can be found in the test script:
https://github.com/ivan-chai/torch-linear-assignment/blob/main/tests/test_to_indices.py
